# Schedule for Session V


## Day 1

| Time            |   Topic  | Instructor |
|:------------------------:|:----------:|:--------:|
|09:00 - 9:30 | [Homework Review]() | All |
|09:30 - 10:15 | [Single-cell RNA-seq - marker identification](https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionIV/lessons/SC_marker_identification.html) | Mary |
|10:15 - 10:30 | Break | |
|10:30 - 11:30 | [Introduction to ChIP-seq](https://github.com/hbctraining/In-depth-NGS-Data-Analysis-Course/blob/master/sessionV/slides/Introduction%20to%20ChIP-seq%20Summer%202018.pdf) | Dr. Shannan Ho Sui |
|11:30 - 12:00 | [Setting up for ChIP-seq analysis](https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionV/lessons/01_Intro_chipseq_data_organization.html) | Meeta |
|12:00 - 13:00 | Lunch | |
|13:00 - 14:15 | [Alignment and filtering of reads](https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionV/lessons/03_align_and_filtering.html) | Mary |
|14:15 - 14:35 | [ChIP-seq File Formats](https://github.com/hbctraining/In-depth-NGS-Data-Analysis-Course/raw/master/sessionV/slides/Workflows_and_fileformats.pdf) | Radhika |
|14:35 - 15:05 | [Peak calling with MACS2](https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionV/lessons/04_peak_calling_macs.html) | Meeta |
|15:05 - 15:15 | Break | |
|15:15 - 16:00 | [Peak calling with MACS2 (contd.)](https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionV/lessons/04_peak_calling_macs.html#running-macs2) | Meeta |
|16:00 - 17:00 | [ChIPQC: generating a report with quality metrics](https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionV/lessons/05_combine_chipQC_and_metrics.html) | Radhika |

## Day 2

| Time            |  Topic  | Instructor |
|:------------------------:|:----------:|:--------:|
|09:00 - 10:15 | [ChIPQC: Assessing ChIP quality](https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionV/lessons/05_combine_chipQC_and_metrics.html) | Radhika |
|10:15 - 10:45 | [Handling Replicates with Bedtools](https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionV/lessons/06_handling-replicates-bedtools.html) | Mary |
|10:45 - 10:55 | Break | |
|10:55 - 11:15 | [Handling Replicates with IDR](https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionV/lessons/07_handling-replicates-idr.html) | Radhika |
|11:15 - 12:15 | [Differentially enriched peaks using DiffBind](https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionV/lessons/08_diffbind_differential_peaks.html) | Meeta |
|12:15 - 13:15 | Lunch | |
|13:15 - 14:30 | [Visualization and exploration of ChIP-seq data](https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionV/lessons/10_data_visualization.html) | Meeta |
|14:30 - 15:20 | [Qualitative assessment using IGV](https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionV/lessons/11_qualitative_assessment_IGV.html) | Radhika |
|15:30 - 15:40 | Break | |
|15:20 - 16:20 | [Functional analysis](https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionV/lessons/12_annotation_functional_analysis.html) | Mary |
|16:20 - 16:40 | [Integrating ChIP-seq data with RNA-seq](https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionV/lessons/integrating_rna-seq_and_chip-seq.html) | Meeta |
|16:40 - 17:00 | [Overview of ChIP-seq workflow](https://github.com/hbctraining/In-depth-NGS-Data-Analysis-Course/blob/master/sessionV/slides/ChIP-seq_troubleshooting.pdf) | Meeta |
